# scWECTA
scWECTA is a soft-weighted ensemble model for single-cell cell-type annotation using five different informative gene sets and integrating 
five classifiers.

## Usage
The workflow of scWECTA mainly contain three steps.<br /> 

(1) Data preprocessing<br />
(2) Model training<br/>
(3) Model predicting<br/>


### Data preprocessing
Suppose your working directory is named as path<br />
  
The data preprocessing is mainly done with preprocess.R. Take scRNA-seq raw count matrix as input (make sure train and test count matrix 
both in above directory), and the command line under linux is as follows,
```
Rscript preprocess.R -ref "path/train_count" -query "path/test_count"
```


### Model training and model predicting
Put the processed data generated by data preprocessing into different directories. <br /> 

Put training data and gene sets generated by the training data into "path/train". Put the test data into "path/test/". Don't forget the 
marker gene list,  in the Data in our github dictory, put it into "path/marker/". Besides that, create a new output file directory named "path/output/"
  
These two steps contain in the scWECTA.py, and user could run it in the linux. We use python 3.7, and the command line is as follows,
```
python scWECTA.py -train "path/train" -test "path/test/" -marker "path/marker/" -o "path/output/" -s "Human" -t "Pancreas" -thred 0.5
```
